ext/git.c

```diff
@@ -0,0 +1,339 @@
+        remove_credentials(repository_url);
+        object_init_ex(carrier, ddtrace_ce_git_metadata);
+        ddtrace_git_metadata *git_metadata = (ddtrace_git_metadata *) Z_OBJ_P(carrier);
+        ZVAL_STR(&git_metadata->property_commit, commit_sha);
+        ZVAL_STR(&git_metadata->property_repository, repository_url);
+        return true;
+    }
+
+    return false;
+}
+
+git_source inject_from_env(zval *carrier) {
+    zend_string *commit_sha = get_DD_GIT_COMMIT_SHA();
+    zend_string *repository_url = get_DD_GIT_REPOSITORY_URL();
+    return add_git_info(carrier, commit_sha, repository_url) ? GIT_SOURCE_ENV : GIT_SOURCE_NONE;
+}
+
+git_source inject_from_global_tags(zval *carrier) {
+    bool success = false;
+    zend_array *global_tags = get_DD_TAGS();
+    zval *commit_sha = zend_hash_str_find(global_tags, ZEND_STRL("git.commit.sha"));
+    zval *repository_url = zend_hash_str_find(global_tags, ZEND_STRL("git.repository_url"));
+    if (commit_sha && repository_url && Z_TYPE_P(commit_sha) == IS_STRING && Z_TYPE_P(repository_url) == IS_STRING) {
+        success = add_git_info(carrier, Z_STR_P(commit_sha), Z_STR_P(repository_url));
+    }
+    return success ? GIT_SOURCE_GLOBAL_TAGS : GIT_SOURCE_NONE;
+}
+
+git_source inject_from_git_dir(zval *carrier, zend_string *cwd) {
+    zend_string *git_dir = find_git_dir(ZSTR_VAL(cwd));
+    if (!git_dir) {
+        return false;
+    }
+
+    zend_string *commit_sha = get_commit_sha(ZSTR_VAL(git_dir));
+    zend_string *repository_url = get_repository_url(ZSTR_VAL(git_dir));
+    bool success = add_git_info(carrier, commit_sha, repository_url);
+    zend_string_release(git_dir);
+
+    return success ? GIT_SOURCE_GIT_DIR : GIT_SOURCE_NONE;
+}
+
+void cache_git_metadata(zend_string *cwd, zend_string *commit_sha, zend_string *repository_url, git_source source) {
+    struct _git_metadata *git_metadata = pemalloc(sizeof(struct _git_metadata), 1);
+    git_metadata->source = source;
+    git_metadata->property_commit = commit_sha;
+    git_metadata->property_repository = repository_url;
+    zend_hash_add_ptr(&DDTRACE_G(git_metadata), cwd, git_metadata);
+}
+
+void inject_git_metadata(zval *carrier, zend_string *cwd) {
+    git_source source = inject_from_env(carrier);
+    if (source == GIT_SOURCE_NONE) {
+        source = inject_from_global_tags(carrier);
+    }
+    if (source == GIT_SOURCE_NONE) {
+        source = inject_from_git_dir(carrier, cwd);
+    }
+
+    if (source != GIT_SOURCE_NONE) {
+        ddtrace_git_metadata *git_metadata_obj = (ddtrace_git_metadata *) Z_OBJ_P(carrier);
+        zend_string *property_commit = Z_STR(git_metadata_obj->property_commit);
+        zend_string *property_repository = Z_STR(git_metadata_obj->property_repository);
+        zend_string_addref(property_commit);
+        zend_string_addref(property_repository);
+        cache_git_metadata(cwd, property_commit, property_repository, source);
```

Comment:
You may only cache if you're extracting from directory. Env and tags may change anytime, and you'd not be able to invalidate it.